pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -0,0 +1,282 @@
+                                 to, ImmutableMap.of(COMPACTED_TOPIC_LEDGER_PROPERTY, ledger.getId())))
+            .whenComplete((res, exception) -> {
+                    if (exception != null) {
+                        deleteLedger(bk, ledger)
+                            .whenComplete((res2, exception2) -> {
+                                    // complete with original exception
+                                    promise.completeExceptionally(exception);
+                                });
+                    } else {
+                        promise.complete(ledger.getId());
+                    }
+                });
+        return promise;
+    }
+
+    private void phaseTwoLoop(RawReader reader, MessageId to, Map<String, MessageId> latestForKey,
+                              LedgerHandle lh, Semaphore outstanding, CompletableFuture<Void> promise) {
+        reader.readNextAsync().whenComplete(
+                (m, exception) -> {
+                    try {
+                        if (exception != null) {
+                            promise.completeExceptionally(exception);
+                            return;
+                        } else if (promise.isDone()) {
+                            return;
+                        }
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+
+                        if (latestForKey.get(key).equals(id)) {
+
+                            outstanding.acquire();
+                            CompletableFuture<Void> addFuture = addToCompactedLedger(lh, m)
+                                .whenComplete((res, exception2) -> {
+                                        outstanding.release();
+                                        if (exception2 != null) {
+                                            promise.completeExceptionally(exception2);
+                                        }
+                                    });
+                            if (to.equals(id)) {
+                                addFuture.whenComplete((res, exception2) -> {
+                                        if (exception2 == null) {
+                                            promise.complete(null);
+                                        }
+                                    });
+                            }
+                        }
+                        phaseTwoLoop(reader, to, latestForKey, lh, outstanding, promise);
+                    } catch (InterruptedException ie) {
+                        Thread.currentThread().interrupt();
+                        promise.completeExceptionally(ie);
+                    } finally {
+                        m.close();
+                    }
+                });
+    }
+
+    private CompletableFuture<LedgerHandle> createLedger(BookKeeper bk) {
+        CompletableFuture<LedgerHandle> bkf = new CompletableFuture<>();
+        bk.asyncCreateLedger(conf.getManagedLedgerDefaultEnsembleSize(),
```

Comment:
Should we get the same level of replication as configured for the namespace?

Reply:
How do I access that? I've added a task to change it to namespace configured.

Reply:
Sure we can add later, in generate through the `PulsarService` instance it's possible to access the configuration and the namespace policies. If the `PulsarService` is not available, we could reuse the code the `ConfigurationCache` class that gets it from global ZK.